### ISSUE TITLE
installing  kernel from virtualenv (on macos) points to system binary

### ISSUE THREAD

**Issue by kubaraczkowski** [ipython/ipython#11569]
Hi,

When installing ipykernel kernelspecs from inside virtualenv I'm expecting to have kernelspec.json to point to the binary of the virtualenv, however, I get the system binary.

This happens on **macOS** with python3 installed via homebrew.
 **python 3.7.1, ipykernel 5.1.0.**

Steps to reproduce:
```
python3 -m venv --system-site-packages ipykerneltest
source ./ipykerneltest/bin/activate
python3 -m ipykernel install --name ipykerneltest --user
head ~/Library/Jupyter/kernels/ipykerneltest/kernel.json
```

I'm getting:
```
{
 "argv": [
  "/usr/local/opt/python/bin/python3.7",
...
```

Am I doing it all wrong, or is it a bug somewhere?

Regards to the team!

**Comment by Carreau** [comment]
You will need to  follow the instruction on [how to manage kernels in multiple environements](https://ipython.readthedocs.io/en/stable/install/kernel_install.html#kernels-for-different-environments)

**Comment by kubaraczkowski** [comment]
I have indeed.

I've figured out though that the behaviour depends whether the virtualenv is created with "`--system-site-packages`" or not.
In an "isolated" venv the ipykernel install points to the correct python executable.
In an "`--system-site-packages`" one, if ipykernel is already present in the "system packages" the executable of the "system" is taken, not of the virtualenv.

Is this the expected behavior?

**Comment by Carreau** [comment]
Apologies for the previous scarce reply, I had to run, and did not completely read your issue.

From the top of my head I'm unsure which ipykernel will take precedence and what's the logic beyond what you see, it does appear weird indeed, but my guess is that we are using heuristics to find which python executable to run based on the site-packages location.

So my wild guess is the `ipykernel install` think "oh system site package so I'm going to use system  python".

I'm afraid thee logic of that might be in `ipykernel` , or even `jupyter_client`, so I'll have to dig deep. @takluyver  and @mpacer worked recently on kernelspec I believe so might know.

**Comment by kubaraczkowski** [comment]
No need for apologies. I also tried restarting my computer before asking ;)

Indeed, it seems like the kernelspec is created as you describe. Correct me
if i'm wrong, but that's not what the user might expect.
The scenario in which we're working is that there's jupyter installed in
the 'system' (brewed) python and then virtualenvs are 'linked' to that
central system. this way the user needs to start one jupyter session only.
That means though that ipykernel is also present in the
'system-site-packages'.

Is it really needed to use heuristics here? sys.executable should normally
be all that's needed I'd guess?

**Comment by kubaraczkowski** [comment]
You're right, i've digged into the kernelspec.py and sys.executable it is!

However, it seems that using --system-site-packages does influence
sys.executable. Not only in ipykernel, but in general. This is kind of
unexpected, but perhaps needed to get the imports right. Interestingly the
sys.base_path and sys.path remain unaffected...

I looked for a reliable way of figuring out the 'effective' executable, but
didn't find one. Maybe you have more experience with that?

**Comment by takluyver** [comment]
Does the venv you create (with `--system-site-packages`) have a `bin/python3` itself? And if so, is it a symlink to your system Python? And does this differ when `--system-site-packages` isn't used?

**Comment by alexgawrilow** [comment]
I had the same problem on Ubuntu.
After installing the ipykernel inside the virtual environment, the python path was set to the standard python path (/usr/bin/python3). However, when I deactivated the virtual environment and reactivated it and installed another ipykernel in the virtual environment, the python path was set correctly to the path of the virtual environment.

**Comment by waiyip-aquabyte** [comment]
@alexgawrilow's workaround does it for me. I have never used the `--system-site-packages` option.
